### PR TITLE
AirfRANS: golden config T_max=5+gc=1.0+WD=1e-2 multi-seed (seeds 100-104)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Multi-seed characterization of the compound golden config (T_max=5 + grad-clip=1.0 + WD=1e-2). Even if kohaku's single-seed run (PR #2741) beats baseline, we need seed variance data to understand the result distribution. Previous multi-seed sweeps revealed 20-30% range between best/worst seeds (PR #2655: seeds 0.0153–0.0194). A lucky seed at this golden config could push well below 0.012 — possibly approaching 0.011 or lower.

**Why 5 seeds:** AirfRANS phase transitions are stochastic — timing varies by seed. T_max=5 produces 8 cosine troughs in 41 epochs, giving 8 chances for a deep phase transition per seed. With 5 seeds × 8 troughs = 40 opportunities to find a deep basin. The best result from this sweep becomes the new anchor for further experiments.

**Why WD=1e-2 + T_max=5:** This combines the two best improvements discovered so far:
- T_max=5 (PR #2732): 0.01271 — more cosine cycles for deeper basin exploration
- WD=1e-2 (PR #2709): 0.01323 — clean L2 regularization with grad-clip preventing spikes

## Instructions

Run **5 separate training runs** with seeds 100–104. Each run uses the same golden config — only `--seed` changes:

```bash
# Seed 100
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed 100 --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed100"

# Seed 101
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed 101 --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed101"

# Seed 102
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed 102 --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed102"

# Seed 103
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed 103 --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed103"

# Seed 104
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed 104 --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed104"
```

**MANDATORY flags:** `--epochs 999` and `--no-use-ema` for all runs.

**What to report (for each seed):**
- Best `val_primary/surface_mse` and the epoch it occurred at
- Whether best epoch is at a cosine trough (multiples of 5) or peak
- W&B run ID

**Summary table format:**
| Seed | Best val_primary/surface_mse | Best Epoch | W&B Run |
|------|------------------------------|------------|---------|
| 100  | X.XXXXX | N | run_id |
| ...  | ... | ... | ... |

Also report: mean, std, min, max across all 5 seeds.

## Baseline

- **Primary metric:** `val_primary/surface_mse` (lower is better)
- **Current best:** 0.01271 (PR #2732, kohaku — T_max=5 + grad-clip=1.0 + WD=1e-4 default, single seed)
- **Baseline W&B run:** uh7fchiy
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999`

**Seed variance reference (PR #2655, lr=3e-4 + T_max=10):**

| Seed | val_primary/surface_mse |
|------|------------------------|
| 789  | 0.0153 (best) |
| 456  | 0.0170 |
| 123  | 0.0182 |
| 42   | 0.0193 |
| 1337 | 0.0194 (worst) |

Range: 0.0153–0.0194 (21% spread). Expect similar or wider spread with higher LR (7e-4).

**External target:** 0.0043 (~3x gap remaining)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Multi-seed sweep: golden config T_max=5 + grad-clip=1.0 + WD=1e-2 (seeds 100–104)

| Seed | Best val_primary/surface_mse | Best Epoch | Cosine Trough (mult of 5)? | W&B Run |
|------|------------------------------|------------|----------------------------|---------|
| 100  | 0.02223 | 39 | NO | jeuyusit |
| 101  | **0.01002** | 37 | NO | hgs11esu |
| 102  | 0.01764 | 35 | YES | eg81dgfb |
| 103  | 0.01066 | 41 | NO | kupl6z2w |
| 104  | 0.01237 | 39 | NO | nlmqw6fn |

### Summary statistics

| Stat | val_primary/surface_mse |
|------|------------------------|
| Mean | 0.01458 |
| Std  | 0.00467 |
| Min  | **0.01002** (seed 101) |
| Max  | 0.02223 (seed 100) |
| Range | 0.01002 – 0.02223 (122% spread) |

### Comparison to baseline

- **Baseline:** 0.01271 (PR #2732, kohaku — T_max=5 + grad-clip=1.0 + WD=1e-4)
- **Seeds beating baseline:** 2/5 (seed 101: 0.01002 = -21.2%, seed 103: 0.01066 = -16.1%)
- **Seeds worse than baseline:** 3/5 (seed 104: 0.01237 is close, seeds 100 and 102 are notably worse)

### Final test metrics (full_test/surface_mse at last epoch)

| Seed | full_test/surface_mse |
|------|----------------------|
| 100  | 0.01979 |
| 101  | 0.03323 |
| 102  | 0.03209 |
| 103  | 0.01322 |
| 104  | 0.02263 |

### Commands used

All runs used identical config except seed:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=<0-4> python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --seed <100-104> --wandb-group airfrans-golden-tmax5-seeds --wandb-name "emma/airfrans-golden-tmax5-seed<100-104>" --agent emma
```

Each run completed 41 epochs (capped by SENPAI_MAX_EPOCHS=50 / SENPAI_TIMEOUT_MINUTES=30). Peak memory: not explicitly logged, but all runs completed without OOM on RTX PRO 6000 Blackwell (96GB).

### What happened

**The WD=1e-2 golden config shows very high seed variance — much wider than previous sweeps.** The 122% spread (0.01002 to 0.02223) is dramatically larger than the 21% spread seen in PR #2655 (seeds at lr=3e-4, T_max=10). This suggests the higher weight decay (1e-2 vs 1e-4) amplifies stochastic sensitivity.

Key observations:
1. **Best result is excellent:** Seed 101 at 0.01002 is a new best — 21% below the baseline of 0.01271. This is the first result breaking below 0.011.
2. **But unreliable:** Only 2/5 seeds beat the baseline. The mean (0.01458) is worse than baseline.
3. **Best epochs are mostly NOT at cosine troughs:** 4/5 seeds peaked at non-trough epochs (37, 39, 41), suggesting the best basins are found during the cosine ramp-up/descent phases, not at the minimum LR points.
4. **Final test ≠ best validation:** The gap between best validation and final test metrics (e.g., seed 101: val 0.01002 vs test 0.03323) shows high epoch-to-epoch volatility in this regime. Early stopping / best-checkpoint selection would be critical for production use.

### Suggested follow-ups

1. **Best-checkpoint selection:** The large val/test gap suggests implementing early stopping or checkpoint-based model selection at the best validation epoch could lock in gains.
2. **Lower WD for stability:** WD=5e-3 or WD=3e-3 might give better mean performance with lower variance while still improving over WD=1e-4.
3. **Extended seeds for the best config:** Running seeds 105–109 could help distinguish whether this variance is inherent to the config or whether we got unlucky on 3 seeds.
4. **Seed 101 as anchor:** The 0.01002 result from seed 101 should be the new anchor. Future experiments targeting AirfRANS should try to match this with different configs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2742 because ADVISOR: Closing — gc=1.0+T_max=5+WD=1e-2 multi-seed best was 0.010020 vs 0.00935 baseline. The gc=1.5 breakthrough (PR #2737) makes all gc=1.0 experiments obsolete. Thank you for the multi-seed data — it confirmed T_max=5 > T_max=10 at gc=1.0, which informed our gc=1.5 experiments..
